sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java

```diff
@@ -1532,26 +1767,47 @@ public DataSource buildDatasource() {
         SnowflakeBasicDataSource basicDataSource = new SnowflakeBasicDataSource();
         basicDataSource.setUrl(buildUrl());
 
-        if (getUsername() != null) {
-          basicDataSource.setUser(getUsername());
+        if (isNotEmpty(getOauthToken())) {
+          basicDataSource.setOauthToken(getOauthToken().get());
+        } else if (isNotEmpty(getUsername()) && getPrivateKey() != null) {
+          basicDataSource.setUser(getUsername().get());
+          basicDataSource.setPrivateKey(getPrivateKey());
+        } else if (isNotEmpty(getUsername())
+            && isNotEmpty(getPrivateKeyPassphrase())
+            && isNotEmpty(getRawPrivateKey())) {
+          PrivateKey privateKey =
+              KeyPairUtils.preparePrivateKey(
+                  getRawPrivateKey().get(), getPrivateKeyPassphrase().get());
+          basicDataSource.setPrivateKey(privateKey);
+          basicDataSource.setUser(getUsername().get());
+
+        } else if (isNotEmpty(getUsername()) && isNotEmpty(getPassword())) {
+          basicDataSource.setUser(getUsername().get());
+          basicDataSource.setPassword(getPassword().get());
+        } else {
+          throw new RuntimeException("Missing credentials values. Please check your credentials");
```

Comment:
populateDisplayData is called at pipeline construction time. Since you're adding support for ValueProviders, you're enabling users to fill in the data for their credentials AFTER construction time. Imagine you're building a template. In that case, none of the credentials will be known.

Therefore, I think you likely don't want to throw this exception.

Reply:
Before using provided credentials we're checking them using `isNotEmpty` method. 

We tested it with dataflow templates by providing credentials from Dataflow UI.

Reply:
LGTM then.